### PR TITLE
[Fix] Replace cached time with system clock in MasterService debug logs

### PR DESCRIPTION
**Pull request by sandeshkr419**
### Description
Cached Time gives imprecise duration in computing multiple computation time since it caches every 200ms and multiple operation in MasterService regarding cluster state creation, update, publish require time frames less than 200ms during debugging. So changing the computation time to consume `System.nanotime()` to get clock time instead of cached time. This is essential when debug logs are enabled - else this is not logged by default.

Before:
```
[2023-05-31T11:38:34,142][DEBUG][o.o.c.s.MasterService    ] [master1] took [0s] to compute cluster state update for [index-aliases]
[2023-05-31T11:38:34,307][DEBUG][o.o.c.s.MasterService    ] [master1] took [0s] to notify listeners on successful publication of cluster state (version: 55109, uuid: Fc58ns6mRHCNuxcvADvfOg) for [index-aliases]
[2023-05-31T11:38:32,416][DEBUG][o.o.c.s.MasterService    ] [master1] took [0s] to compute cluster state update for [cluster_reroute(reroute after starting shards)]
[2023-05-31T11:38:32,416][DEBUG][o.o.c.s.MasterService    ] [master1] took [0s] to notify listeners on unchanged cluster state for [cluster_reroute(reroute after starting shards)]
[2023-05-31T11:38:34,307][DEBUG][o.o.c.s.MasterService    ] [master1] took [0s] to notify listeners on successful publication of cluster state (version: 55109, uuid: Fc58ns6mRHCNuxcvADvfOg) for [index-aliases]
```

After:
```
[2023-06-02T12:28:49,551][DEBUG][o.o.c.s.MasterService    ] [master1] took [2ms] to notify listeners on unchanged cluster state for [index-aliases]
[2023-06-02T13:10:17,855][DEBUG][o.o.c.s.MasterService    ] [master1] took [93ms] to compute cluster state update for [index-aliases]
[2023-06-02T13:10:47,875][DEBUG][o.o.c.s.MasterService    ] [master1] executing cluster state update for [index-aliases]
[2023-06-02T13:10:47,876][DEBUG][o.o.c.s.MasterService    ] [master1] took [0s] to compute cluster state update for [index-aliases]
[2023-06-02T13:10:47,877][DEBUG][o.o.c.s.MasterService    ] [master1] took [0s] to notify listeners on unchanged cluster state for [index-aliases]
[2023-06-02T13:34:24,247][DEBUG][o.o.c.s.MasterService    ] [master1] took [0s] to notify listeners on successful publication of cluster state (version: 60931, uuid: G8fs82mZROGv4C0ufnehKA) for [index-aliases]
```

### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/issues/7849

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by andrross** [comment]
Looks like some tests are failing?

File: CHANGELOG.md [hunk 1]
```diff
@@ -94,6 +94,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 ### Changed
 - Replace jboss-annotations-api_1.2_spec with jakarta.annotation-api ([#7836](https://github.com/opensearch-project/OpenSearch/pull/7836))
 - Add min, max, average and thread info to resource stats in tasks API ([#7673](https://github.com/opensearch-project/OpenSearch/pull/7673))
+- MasterService Debug logs for computation time to use system clock time instead of cached time ([#7902](https://github.com/opensearch-project/OpenSearch/pull/7902))
```
Comment (andrross) [inline CHANGELOG.md:97]: I'd rephrase this as "Compute accurate elapsed time in MasterService logging output"

**Comment by shwetathareja** [comment]
We should change ```ClusterAplplierService``` class as well.

https://github.com/opensearch-project/OpenSearch/blob/aa2158555c7fea440d8de80b939a8c646aa5d415/server/src/main/java/org/opensearch/cluster/service/ClusterApplierService.java#L715

**Comment by sandeshkr419** [comment]
> I would want us to double click on this to see if we really need this accuracy. Based on my understanding we are simply using this for debugging. What are the actionable items if these numbers stack around few seconds lets say.

@Bukhtawar Given the frequency of these logging/actions, we should not be causing any regression in performance. Since the default cluster manager election timeout duration is 500ms, majority of these Cluster State updates are expected to be of the order of ms only. 

We might be over-precise in time computation for actions spanning over seconds, but that should be okay if it is not causing any performance regression. Thoughts?

File: CHANGELOG.md (hunk 1 above)
Comment (dblock) [reply CHANGELOG.md:97]: It's really a fix, no? In which case it should be "Correct system clock time in MasterService debug output" and go into fixes.

**Comment by dblock** [comment]
IMHO we should show correct time over incorrect time unless it was a performance optimization, so +1 for this change.

File: CHANGELOG.md (hunk 1 above)
Comment (andrross) [reply CHANGELOG.md:97]: "cached" time was used as an optimization but resulted in 200ms precision/granularity in the elapsed time reported in the logs. One could argue that was a deliberate tradeoff, but calling it a fix is probably simpler.

**Comment by dblock** [comment]
@sandeshkr419 either way, could you please help iterate this to 🟢 ? thanks!

**Comment by sandeshkr419** [comment]
> We should change `ClusterApplierService` class as well.
> 
> https://github.com/opensearch-project/OpenSearch/blob/aa2158555c7fea440d8de80b939a8c646aa5d415/server/src/main/java/org/opensearch/cluster/service/ClusterApplierService.java#L715

@shwetathareja Will check what are the time intervals here to see if we need to change here or not.

**Comment by shwetathareja** [comment]
> @shwetathareja Will check what are the time intervals here to see if we need to change here or not.

Applier thread is run after new state is published. Any new state has to be applied by the node for it to be usable. If any applier/ listener takes longer, it can cause over state application to take long. This also runs in single thread. There are multiple appliers and listeners that run in sequential order. If cluster state application takes in secs, we need accurate break down across those appliers and listeners. I am fine if you want to raise different PR. Please don't close the issue  https://github.com/opensearch-project/OpenSearch/issues/7849 though.

File: server/src/main/java/org/opensearch/cluster/service/MasterService.java [hunk 2]
```diff
@@ -131,7 +132,13 @@ public class MasterService extends AbstractLifecycleComponent {
     protected final ClusterManagerTaskThrottler clusterManagerTaskThrottler;
     private final ClusterManagerThrottlingStats throttlingStats;
 
+    private final LongSupplier nanoTimeSupplier;
+
     public MasterService(Settings settings, ClusterSettings clusterSettings, ThreadPool threadPool) {
+        this(settings, clusterSettings, threadPool, System::nanoTime);
+    }
+
+    public MasterService(Settings settings, ClusterSettings clusterSettings, ThreadPool threadPool, LongSupplier timeProvider) {
```
Comment (shwetathareja) [inline server/src/main/java/org/opensearch/cluster/service/MasterService.java:141]: keep variable name same across classes - timeProvider/ timeSupplier

File: server/src/main/java/org/opensearch/cluster/service/ClusterManagerService.java [hunk 3]
```diff
@@ -21,4 +23,8 @@ public class ClusterManagerService extends MasterService {
     public ClusterManagerService(Settings settings, ClusterSettings clusterSettings, ThreadPool threadPool) {
         super(settings, clusterSettings, threadPool);
     }
+
+    public ClusterManagerService(Settings settings, ClusterSettings clusterSettings, ThreadPool threadPool, LongSupplier nanoTimeSupplier) {
```
Comment (shwetathareja) [inline server/src/main/java/org/opensearch/cluster/service/ClusterManagerService.java:27]: can you add the caller which will call this new constructor

Comment (shwetathareja) [reply server/src/main/java/org/opensearch/cluster/service/ClusterManagerService.java:27]: why do we need supplier, any particular reason?

File: server/src/main/java/org/opensearch/cluster/service/MasterService.java (hunk 2 above)
Comment (sandeshkr419) [reply server/src/main/java/org/opensearch/cluster/service/MasterService.java:141]: Missed during final refactoring, will revise it in next commit.

File: server/src/main/java/org/opensearch/cluster/service/ClusterManagerService.java (hunk 3 above)
Comment (sandeshkr419) [reply server/src/main/java/org/opensearch/cluster/service/ClusterManagerService.java:27]: > why do we need supplier, any particular reason?

1. Better abstraction - don't need to scavenge through entire class to make changes in future if required.
2. Ease of testing - else we will have to go override the private methods (ofc, by declaring them protected just for testing) in tests like for [ClusterApplierService](https://github.com/opensearch-project/OpenSearch/blob/aa2158555c7fea440d8de80b939a8c646aa5d415/server/src/main/java/org/opensearch/cluster/service/ClusterApplierService.java#L714) in [ClusterApplierServiceTests](https://github.com/opensearch-project/OpenSearch/blob/aa2158555c7fea440d8de80b939a8c646aa5d415/server/src/test/java/org/opensearch/cluster/service/ClusterApplierServiceTests.java#L626). So having a Supplier seemed a clean design choice.

File: server/src/main/java/org/opensearch/cluster/service/MasterService.java (hunk 2 above)
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/cluster/service/MasterService.java:141]: Should threadpool itself have a method to fetch non-cached time to prevent overhauling of code at multiple places. Then the cached time is 200ms by default we might want to see how system performs with lower cache times 100ms, 50ms etc and would that make sense for other use cases to switch to lower cache time or non-cached times alike

File: server/src/test/java/org/opensearch/cluster/service/MasterServiceTests.java [hunk 4]
```diff
@@ -97,16 +98,15 @@
 public class MasterServiceTests extends OpenSearchTestCase {
 
     private static ThreadPool threadPool;
-    private static long relativeTimeInMillis;
+    private static Long relativeTimeInMillis;
+
+    private static LongSupplier nanoTimeSupplier() {
+        return () -> relativeTimeInMillis * 1000000L;
```
Comment (dbwiddis) [inline server/src/test/java/org/opensearch/cluster/service/MasterServiceTests.java:104]: Would prefer a constant here rather than a magic number, e.g. NANOS_PER_MILLIS.  I'm also fond of using underscores to make it easy to count zeroes.  Less important as this is a test class.

File: server/src/main/java/org/opensearch/cluster/service/MasterService.java (hunk 2 above)
Comment (shwetathareja) [reply server/src/main/java/org/opensearch/cluster/service/MasterService.java:141]: > Should threadpool itself have a method to fetch non-cached time to prevent overhauling of code at multiple places.

+1

I dont like adding new constructor for ClusterManagerService class to just pass the timeProviderSupplier

Comment (sandeshkr419) [reply server/src/main/java/org/opensearch/cluster/service/MasterService.java:141]: Seems okay, we might have to overload the ThreadPool's new time provider function for tests like we do in existing code. Will draft the changes with this suggestion.

File: server/src/test/java/org/opensearch/cluster/service/MasterServiceTests.java (hunk 4 above)
Comment (sandeshkr419) [reply server/src/test/java/org/opensearch/cluster/service/MasterServiceTests.java:104]: Thanks, I guess we might be moving away with this change to have a method in ThreadPool, but I'll account this if applicable in further changes.

File: server/src/main/java/org/opensearch/cluster/service/MasterService.java (hunk 2 above)
Comment (sandeshkr419) [reply server/src/main/java/org/opensearch/cluster/service/MasterService.java]: @Bukhtawar @shwetathareja @andrross Iterated the PR based on having a method inside ThreadPool itself to fetch absolute time. Please review now.

@shwetathareja I haven't included [ClusterApplierService](https://github.com/opensearch-project/OpenSearch/blob/aa2158555c7fea440d8de80b939a8c646aa5d415/server/src/main/java/org/opensearch/cluster/service/ClusterApplierService.java#L714) changes here now to avoid rework if further iterations are required. Will raise a separate PR once this is merged.

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java [hunk 5]
```diff
@@ -330,6 +330,15 @@ public long relativeTimeInNanos() {
...
 
+    /**
+     * Returns a value of nanoseconds that may be used for absolute time calculations.
+     *
+     * This method can be used for calculating precise time deltas.
+     */
+    public long absoluteTimeInNanos() {
+        return System.nanoTime();
+    }
+
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/threadpool/ThreadPool.java:341]: The name is misleading. Absolute time generally refers to wall clock time System#currentTimeMillis vs relative time is the logical time System#nanoTime

Comment (sandeshkr419) [reply server/src/main/java/org/opensearch/threadpool/ThreadPool.java:341]: I'm open to suggestions, how about `getNanoTime` or `getSystemNanoTime` instead of `absoluteTimeInNanos`? Don't want to use `relative` word here since then the concern will be relative to what.

File: server/src/main/java/org/opensearch/cluster/service/ClusterManagerService.java (hunk 3 above)
Comment (sandeshkr419) [reply server/src/main/java/org/opensearch/cluster/service/ClusterManagerService.java]: Resolving this conversation since we changed the implementation entirely to have a utility method inside `ThreadPool` class itself.

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java (hunk 5 above)
Comment (andrross) [reply server/src/main/java/org/opensearch/threadpool/ThreadPool.java:341]: System.nanoTime() gives you the number of nanoseconds since an unknown and arbitrary point. Therefore it is "relative" in that it is only useful when you compare it relative to another invocation to System.nanoTime in the same JVM instance. So "relative" does make sense here, but the challenge is to differentiate it from the existing "relativeTime..." methods.

I might suggest something simple like "uncachedRelativeTimeInNanos" but I'd be happy to hear other suggests. Make sure the javadoc gives some guidance as to when to use this versus the other "relativeTime..." methods.

Comment (sandeshkr419) [reply server/src/main/java/org/opensearch/threadpool/ThreadPool.java:341]: @andrross @Bukhtawar Thanks for the suggestions. Putting more thoughts through, I'd prefer `getSystemNanoTime` to simply state that it's a wrapper over `System.nanoTime()`. Anything else in the method name will demand extra information to be conveyed through javadoc and method definition.

While javadoc should give guidance on "when" to use this method versus other relative time, the name should be clear and convey "what" it does without having to look through function definition / javadoc.

Thoughts?

Comment (andrross) [reply server/src/main/java/org/opensearch/threadpool/ThreadPool.java:341]: "what" it does is give you a precise relative time. System.nanoTime is "how" it does it, and the caller should not need to be concerned with the "how". What do you think of the following?

```
/**
 * Returns a value of nanoseconds that may be used for relative time calculations
 * that require the highest precision possible. Performance critical code must use
 * either {@link #relativeTimeInNanos()} or {@link #relativeTimeInMillis()} which
 * give better performance at the cost of lower precision.
 *
 * This method should only be used for calculating time deltas. For an epoch based
 * timestamp, see {@link #absoluteTimeInMillis()}.
 */
public long preciseRelativeTimeInNanos() {
    return System.nanoTime();
}
```

**Comment by dblock** [comment]
@andrross want to take a look at the updated version? much appreciated

File: server/src/main/java/org/opensearch/threadpool/ThreadPool.java (hunk 5 above)
Comment (sandeshkr419) [reply server/src/main/java/org/opensearch/threadpool/ThreadPool.java]: @andrross I guess your latest suggestion makes sense. I guess thumbs up by @Bukhtawar & @shwetathareja shows their agreement as well. I've refactored accordingly.

**Review by shwetathareja — APPROVED** [review]

**Review by andrross — APPROVED** [review]
